ksql-engine/src/main/java/io/confluent/ksql/planner/plan/AggregateNode.java

```diff
@@ -263,15 +278,29 @@ public WindowExpression getWindowExpression() {
         aggregationContext
     );
 
-    if (havingExpressions != null) {
+    final Optional<Expression> havingExpression = Optional.ofNullable(havingExpressions)
+        .map(internalSchema::resolveToInternal);
+
+    if (havingExpression.isPresent()) {
       aggregated = aggregated.filter(
-          internalSchema.resolveToInternal(havingExpressions),
+          havingExpression.get(),
           contextStacker.push(FILTER_OP_NAME),
           builder.getProcessingLogContext());
     }
 
+    final List<SelectExpression> finalSelects = internalSchema
+        .updateFinalSelectExpressions(getFinalSelectExpressions());
+
+    materializationInfo = Optional.of(MaterializationInfo.of(
+        AGGREGATE_STATE_STORE_NAME,
+        aggregationSchema,
+        havingExpression,
+        schema,
+        finalSelects
```

Comment:
By definition the final selects can't already be already applied to the state store.  They're job is to select (and rename) the fields from the statestore schema needed to build the output.

consider:

`SELECT  COUNT() * 3, COUNT() - 1 FROM X GROUP BY Y;`

The aggregate schema will store the value for `COUNT()` and the final selects will apply the transformation required to build the two output fields from that value.

I don't think we'd want to store `COUNT() * 3, COUNT() - 1` as two separate fields in the state store, and there are other examples where I don't even think that would work.

And yes #3302 makes this even more complicated :(

Reply:
yes indeed! That's what I meant by "the only way to make it possible" - it just seems like it may cause issues with the "duplicated" logic down the line.

Reply:
I don't think there'll be much duplication. The actual conversion is handled by the select mapper, which is common across both use cases.  We'll end up with different physical steps at some point - one for handling projections for classic KS topologies, and one for handling them against materialized tables.  However, that's unavoidable - we can't use the same for both, and they'll both internally use the same mapper.